torch/sparse/triton_ops/triton_bsr_dense_mm.py

```diff
@@ -0,0 +1,241 @@
+
+    # Introduce fake batch dimension if not present for convenience.
+    crow_indices = unsqueeze_batch_dim(bsr.crow_indices(), 1)
+    col_indices = unsqueeze_batch_dim(bsr.col_indices(), 1)
+    values = unsqueeze_batch_dim(bsr.values(), 3)
+    dense = unsqueeze_batch_dim(dense, 2)
+    nnz = values.shape[-3]
+    blocksize = values.shape[-2:]
+
+    # Compute broadcasted batch dimension
+    bsr_batch_dims = values.shape[:-3]
+    dense_batch_dims = dense.shape[:-2]
+    batch_dims_broadcasted = torch.broadcast_shapes(bsr_batch_dims, dense_batch_dims)
+    # Required to undo the fake batch dimension insertion.
+    original_batch_dims_broadcasted = torch.broadcast_shapes(
+        bsr.shape[:-2],
+        dense.shape[:-2]
+    )
+
+    # Allocate output
+    output = torch.zeros(batch_dims_broadcasted + (m, n), dtype=dense.dtype, device=dense.device)
+
+    # Broadcast batch dimensions and squash
+    def batch_broadcast_and_squash(t, batch_dims, invariant_dims):
+        return t.broadcast_to(batch_dims + invariant_dims).flatten(0, len(batch_dims) - 1)
+
+    crow_indices = batch_broadcast_and_squash(crow_indices, batch_dims_broadcasted, (-1,))
+    col_indices = batch_broadcast_and_squash(col_indices, batch_dims_broadcasted, (-1,))
+    values = batch_broadcast_and_squash(values, batch_dims_broadcasted, values.shape[-3:])
+    dense = batch_broadcast_and_squash(dense, batch_dims_broadcasted, dense.shape[-2:])
+    # NOTE: output is contiguous, so batch_broadcast_and_squash will create a view
+    output = batch_broadcast_and_squash(output, batch_dims_broadcasted, output.shape[-2:])
+
+    # NOTE: this function will ALWAYS create a view
+    def tile_to_blocksize(t, blocksize):
+        *rest, m, n = t.shape
+        new_shape = rest + [m // blocksize[0], blocksize[0], n // blocksize[1], blocksize[1]]
+        return t.reshape(new_shape).transpose(-3, -2)
+
+    # "Blockify" the row dimension of dense with blocksize[1]
+    # since dense is on the rhs of matmul
+    dense = tile_to_blocksize(dense, blocksize[::-1])
+    # "Blockify" the row dimension of output with blocksize[0]
+    # which is inherited from the bsr input.
+    # NOTE: tile_to_blocksize will create a view.
+    # NOTE: output.blocksize[-1] == dense.blocksize[-1],
+    # so it could be any value in [1, dense.shape[-1]).
+    # We need to probably use the largest possible blocksize
+    # so that it fits into SRAM.
+    output = tile_to_blocksize(output, (blocksize[0], blocksize[0]))
+
+    # Compute a vector of non-zero elements numbers per each row.
+    # We want to ultimately iterate over non-zero rows.
+    nnz_per_row = crow_indices[:, 1:] - crow_indices[:, :-1]
+
+    # Compute indices of non-zero counts.
+    # batch_idx maps to a broadcasted batch index, while
+    # row_idx tracks non-zero rows of the sparse argument
+    # and rows of the output that get modified.
+    batch_idx, row_idx = nnz_per_row.nonzero(as_tuple=True)
+
```

Comment:
`nonzero` could be removed when the number of nonzero rows is high, then skipping empty rows could be delegated to the kernel.

Reply:
nonzero is also one of those ops that will cause a sync if I'm not mistaken.

Reply:
Exactly, so it makes sense to avoid using it in some circumstances which also addresses your comments below.